Copy disjoint pool from UR repo

ldorau opened the pull request.

igchor commented on src/pool/disjoint/pool_disjoint_config_parser.hpp:
```diff
@@ -0,0 +1,53 @@
+
+#include <string>
+#include <unordered_map>
+
```
Hm, actually, I think we should not move the pool_disjoint_config_parser.* files.

The parseDisjointPoolConfig is called by UR, and if this is implemented here, we would have to expose this function in the API, which I think we don't want. 

@pbalcer @bratpiorka what do you think? Should we keep those files in the UR?

ldorau replied:
Removed

igchor commented:
From which commit is the code taken? I think there are some changes missing, e.g. https://github.com/oneapi-src/unified-runtime/commit/050223799e78743607ef8b5d94563256ea4aa0f3

igchor commented on src/pool/disjoint/pool_disjoint_heap_manager.cpp:
```diff
@@ -0,0 +1,112 @@
+
+    pool_data->mem_provider = providers[0];
+    pool_data->disjoint_pool = new usm::DisjointPool();
+    pool_data->disjoint_pool->initialize(providers, numProviders, config);
```
Error handling?

ldorau replied:
Done

ldorau replied:
https://github.com/oneapi-src/unified-memory-framework/pull/17

igchor commented on src/pool/disjoint/pool_disjoint_heap_manager.cpp:
```diff
@@ -0,0 +1,112 @@
+
+enum umf_result_t
+disjoint_pool_initialize(umf_memory_provider_handle_t *providers,
+                         size_t numProviders, void *params, void **pool) {
```
Since you are using `new` inside the function you should wrap the function with `try/catch` and translate the exception to a umf error: you can put try catch around the entire function like this:

```
disjoint_pool_initialize(umf_memory_provider_handle_t *providers,
                         size_t numProviders, void *params, void **pool)  try {
  ...
} catch(std::bad_alloc&) {
   return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
} catch (...) {
   return UMF_RESULT_UNKNOWN_ERROR:
}

```

ldorau replied:
Done: https://github.com/oneapi-src/unified-memory-framework/pull/17

igchor commented on src/pool/disjoint/pool_disjoint_heap_manager.cpp:
```diff
@@ -0,0 +1,112 @@
+            sizeof(struct disjoint_memory_pool));
+    if (!pool_data) {
+        fprintf(stderr, "cannot allocate memory for metadata\n");
+        abort();
```
Why abort? Just return an error.

ldorau replied:
Done: https://github.com/oneapi-src/unified-memory-framework/pull/17

igchor commented on src/pool/disjoint/pool_disjoint_heap_manager.cpp:
```diff
@@ -0,0 +1,112 @@
+
+    *pool = (void *)pool_data;
+
+    (void)params;
```
nit: is this needed?

ldorau replied:
Removed: https://github.com/oneapi-src/unified-memory-framework/pull/17

igchor commented on src/pool/disjoint/pool_disjoint_heap_manager.cpp:
```diff
@@ -0,0 +1,112 @@
+    config.PoolTrace = pub_params->PoolTrace;
+
+    struct disjoint_memory_pool *pool_data =
+        (struct disjoint_memory_pool *)malloc(
```
nit: you can use `new` here as well, so it will be consistent with disjoint_pool allocation (and you dont' have to remember which pointer to `free` and which to `delete`)

ldorau replied:
Done: https://github.com/oneapi-src/unified-memory-framework/pull/17

igchor commented on include/pool/pool_disjoint.h:
```diff
@@ -0,0 +1,39 @@
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+
+struct umf_disjoint_pool_params {
```
We're missing `SharedLimits` here (ref: https://github.com/oneapi-src/unified-memory-framework/pull/4/files#diff-5a84664376e1495e272ca9c9c061388acc9c68940fb5aa9b66e8d9a0c84cc061R36). I'm not sure how to best expose it in C though. The idea behind `SharedLimits` is that it is shared between multiple instances of DisjointPool (.e.g for different types of USM).

We don't have to add this in this PR but let's keep in mind.

igchor commented on src/CMakeLists.txt:
```diff
@@ -8,6 +8,9 @@ set(UMF_SOURCES
+    pool/disjoint/pool_disjoint_config_parser.cpp
+    pool/disjoint/pool_disjoint_heap_manager.cpp
     provider/provider_null.c
     provider/provider_trace.c
```
Unrelated to this PR but why do we have pool/provider trace in the public APIs (include/pool and include/provider)? Those were supposed to be for testing only, I don't think we want to expose it to the user

ldorau replied:
It was suggested by @bratpiorka . @bratpiorka Could you explain it to @igchor ?

bratpiorka replied:
Ok we could move them to the test folder, but in the separate PR

ldorau commented:
> From which commit is the code taken? I think there are some changes missing, e.g. [oneapi-src/unified-runtime@0502237](https://github.com/oneapi-src/unified-runtime/commit/050223799e78743607ef8b5d94563256ea4aa0f3)

Updated from [adb8d9e8 ](https://github.com/oneapi-src/unified-runtime/commit/adb8d9e8) to [a1fbbde7](https://github.com/oneapi-src/unified-runtime/commit/a1fbbde7).

@igchor  I just moved the code from the UR repo with only necessary changes (so that it compiles). Can we make any other changes in a separate PR?

igchor commented:
> > From which commit is the code taken? I think there are some changes missing, e.g. [oneapi-src/unified-runtime@0502237](https://github.com/oneapi-src/unified-runtime/commit/050223799e78743607ef8b5d94563256ea4aa0f3)
> 
> Updated from [adb8d9e8 ](https://github.com/oneapi-src/unified-runtime/commit/adb8d9e8) to [a1fbbde7](https://github.com/oneapi-src/unified-runtime/commit/a1fbbde7).
> 
> @igchor I just moved the code from the UR repo with only necessary changes (so that it compiles). Can we make any other changes in a separate PR?

Sure, that's OK. I would only suggest to remove disjoint/pool_disjoint_config_parser.*pp from this PR. We can always add it later if we decide to do so.

bratpiorka commented:
> > > From which commit is the code taken? I think there are some changes missing, e.g. [oneapi-src/unified-runtime@0502237](https://github.com/oneapi-src/unified-runtime/commit/050223799e78743607ef8b5d94563256ea4aa0f3)
> > 
> > 
> > Updated from [adb8d9e8 ](https://github.com/oneapi-src/unified-runtime/commit/adb8d9e8) to [a1fbbde7](https://github.com/oneapi-src/unified-runtime/commit/a1fbbde7).
> > @igchor I just moved the code from the UR repo with only necessary changes (so that it compiles). Can we make any other changes in a separate PR?
> 
> Sure, that's OK. I would only suggest to remove disjoint/pool_disjoint_config_parser.*pp from this PR. We can always add it later if we decide to do so. @bratpiorka unless this is needed for oneccl?

No, this is not needed for oneCCL. @ldorau please remove these files for now.

ldorau commented:
> > Sure, that's OK. I would only suggest to remove disjoint/pool_disjoint_config_parser.*pp from this PR. We can always add it later if we decide to do so. @bratpiorka unless this is needed for oneccl?
> 
> No, this is not needed for oneCCL. @ldorau please remove these files for now.

@bratpiorka @igchor Done
